### PR TITLE
Fixed extraction of a single message from the buffer

### PR DESCRIPTION
**Pull request by saarnold**
Before the entire buffer was returned while only the first message was parsed,
potentially dropping other messages in the buffer.

### CODE REVIEW

File: src/Driver.cpp [hunk 1]
```diff
@@ -28,10 +30,15 @@ void Driver::open(std::string const& uri)
     openURI(uri);
 }
 
-void Driver::read()
+bool Driver::read()
 {
-    readPacket(&buffer[0], buffer.size());
-    parseMessage(&buffer[4],buffer.size()-4);
+    int msg_size = readPacket(&buffer[0], buffer.size());
+    if(msg_size == KVH_1750_MSG_SIZE)
```
Comment (doudou) [inline src/Driver.cpp]: How can one have a non-KVH_1750_MSG_SIZE packet after packet extraction ? `readPacket` throws if no packet has been found.

Comment (saarnold) [reply src/Driver.cpp]: Should not happen yes. But what is the harm in checking return values?

Comment (doudou) [reply src/Driver.cpp]: > But what is the harm in checking return values?

It basically shows that you don't understand what contracts you need to have enforced in your codebase, a.k.a. that you don't understand the codebase ("not understanding" is fine, defending your lack of understanding is not). After a few hundred of "no harm in testing X" you basically make everyone's life a living hell - including yours a few months from now. As a direct example, here, you return a boolean on a method that can only return `true`. And then test for it in a component, creating two dead paths.

Moreover, it shows that the analysis you did for a bug (with the commit messages I've seen, I am assuming that you're trying to fix a bug) is wrong, since the "fix" most probably does not change anything.

Contracts (implicit or explicit ones) are the bread and butter of what makes large-scale software development viable. If you're afraid of having a bug in `extractPacket`, thoroughly unit test it.

Comment (saarnold) [reply src/Driver.cpp]: > It basically shows that you don't understand what contracts you need to have enforced in your codebase, a.k.a. that you don't understand the codebase ("not understanding" is fine, defending your lack of understanding is not). After a few hundred of "no harm in testing X" you basically make everyone's life a living hell - including yours a few months from now. As a direct example, here, you return a boolean on a method that can only return true. And then test for it in a component, creating two dead paths.

It wasn't my intention to create any dead paths. What we have here is a method blindly parsing the given buffer and a method filling the buffer and returning it's size. If we want to enforce any contracts in such cases, there should be at least a comment explaining it, otherwise everyone who looks into this code has to check if this is not a possible point of failure.

> Moreover, it shows that the analysis you did for a bug (with the commit messages I've seen, I am assuming that you're trying to fix a bug) is wrong, since the "fix" most probably does not change anything.

The bug is that extractPacket returns the whole buffer, what I describe in the first comment of this PR.

> If you're afraid of having a bug in extractPacket, thoroughly unit test it.

Sure. I can add a unit test.

Comment (doudou) [reply src/Driver.cpp]: >  What we have here is a method blindly parsing the given buffer and a method filling the buffer and returning it's size. If we want to enforce any contracts in such cases, there should be at least a comment explaining it, 

The contract is in the `iodrivers_base` API. Adding a comment if that is unclear to you is of course a great idea, but testing in the live codepath for a failure of extractPacket is definitely not a good practice. readPacket will only return a packet as sized by extractPacket by contract.

> The bug is that extractPacket returns the whole buffer, what I describe in the first comment of this PR.

This is not how I interpreted the PR message. This is even worse practice. You don't change a public API to check whether a regression has been fixed ...

Basically what you do in this PR is fix a bug, and then test *in the live code* whether the bug has been fixed, leaking that fact to your callers. It might not have been your intent, but you not only did create a dead path, you basically created a path that would be taken only if the PR *did not fix the bug it's attempting to fix*.

Unit tests is *definitely* the right choice here.

Comment (saarnold) [reply src/Driver.cpp]: > The contract is in the iodrivers_base API. Adding a comment if that is unclear to you is of course a great idea, but testing in the live codepath for a failure of extractPacket is definitely not a good practice. readPacket will only return a packet as sized by extractPacket by contract.

Yes, I will leave a comment just for me.

> Basically what you do in this PR is fix a bug, and then test in the live code whether the bug has been fixed, leaking that fact to your callers. It might not have been your intent, but you not only did create a dead path, you basically created a path that would be taken only if the PR did not fix the bug it's attempting to fix.

No, this line was not supposed to check for a bug that was fixed in the same commit.
I agree that the check, due to the contract in the parent class, is not necessary here.

**Comment by saarnold** [comment]
Ping

**Comment by doudou** [comment]
I'm not part of the org anymore, so can't merge.

But :+1: to me
